File: Protogen/protogen/src/core/src/test/resources/over/cpp/Q_c_executePackage.cpp
#include <string.h>
#include <stdlib.h>
#include "Q_c_executePackage.h"
#include "ptools/PackageBufferWriter.h"
#include "ptools/PackageBufferReader.h"

using namespace protocol;

Q_c_executePackage::Q_c_executePackage(\
		uint64_t a_statement_id,\
		struct Statement_flagsMap a_flags,\
		uint32_t a_paramsCnt,\
		int64_t *a_params
		)
{
	statement_id=a_statement_id;
	flags=a_flags;
	paramsCnt=a_paramsCnt;
	params=a_params;
}

Q_c_executePackage::Q_c_executePackage()
{
	statement_id=0;
	memset(&flags,0,sizeof(flags));
	paramsCnt=0;
	params=0;
}

Q_c_executePackage::~Q_c_executePackage()
{
}

void Q_c_executePackage::serializeW(PackageBufferWriter *writer)
{
	Package::serializeW(writer);
	writer->writeUint64(statement_id);
	uint64_t tmp_flags;
	Statement_flagsFactory::getValueByEnumMap(flags, tmp_flags);
	writer->writeUint64(tmp_flags);
	writer->writeUint32(paramsCnt);
	for(int i=0; i<paramsCnt; i++)
	{
		writer->writeVaruint(params[i]);
	}
}

bool Q_c_executePackage::deserializeR(PackageBufferReader *reader)
{
	if (!Package::deserializeR(reader)) return false;
	statement_id=reader->readUint64();
	Statement_flagsFactory::getEnumMapByValue(reader->readUint64(),flags);
	paramsCnt=reader->readUint32();
	for(int i=0; i<paramsCnt; i++)
	{
		bool b; 
		params[i]=reader->readVaruint(b);
	}
	return true;
}

bool Q_c_executePackage::equals(Package *_package)
{
	if (!_package) return false;
	if (_package->getPackageType()!=this->getPackageType()) return false;
	if (!Package::equals(_package)) return false;
	Q_c_executePackage* p=(Q_c_executePackage*)_package;
	if (!(getStatement_id()==p->getStatement_id())) return false;
	if (!(Statement_flagsFactory::mapEquals(getFlags(),p->getFlags()))) return false;
	if (!(getParamsCnt()==p->getParamsCnt())) return false;
	for(int i=0; i<paramsCnt; i++)
		if (!(getParams()[i]==p->getParams()[i]))
			return false;
	return true;
}

File: Protogen/protogen/example/over/java/Q_c_executePackage.java
package pl.edu.mimuw.loxim.protocol.packages;

import java.math.BigInteger;
import java.util.EnumSet;
import pl.edu.mimuw.loxim.protogen.lang.java.template.ptools.Package;
import pl.edu.mimuw.loxim.protogen.lang.java.template.exception.ProtocolException;
import pl.edu.mimuw.loxim.protogen.lang.java.template.streams.PackageInputStreamReader;
import pl.edu.mimuw.loxim.protogen.lang.java.template.streams.PackageOutputStreamWriter;

import pl.edu.mimuw.loxim.protocol.enums.*;

public class Q_c_executePackage extends Package {
	public final static short ID = (short) 66;

	/**
	 **/
	private java.math.BigInteger statement_id=BigInteger.ZERO;
	/**
	 **/
	private EnumSet<Statement_flagsEnum> flags=EnumSet.noneOf(Statement_flagsEnum.class);
	/**
	 **/
	private long paramsCnt=0;
	/**
	 **/
	private Long[] params=new Long[0];

	public Q_c_executePackage() {
	}

	public Q_c_executePackage(java.math.BigInteger a_statement_id,
			EnumSet<Statement_flagsEnum> a_flags, long a_paramsCnt,
			Long[] a_params) {
		statement_id = a_statement_id;
		flags = a_flags;
		paramsCnt = a_paramsCnt;
		params = a_params;

	}

	@Override
	public long getPackageType() {
		return (long) ID;
	}

	@Override
	protected void deserializeW(PackageInputStreamReader reader)
			throws ProtocolException {
		super.deserializeW(reader);

		try {
			statement_id = reader.readUint64();
			java.math.BigInteger tmp_flags = reader.readUint64();
			flags = Statement_flagsEnum.createEnumMapByValue(tmp_flags);
			paramsCnt = reader.readUint32();
			params = new Long[(int)paramsCnt];
			for (int i = 0; i < paramsCnt; i++)
				params[i] = reader.readVaruint();
		} catch (Exception e) {
			throw new ProtocolException(e);
		}
	}

	@Override
	protected void serializeW(PackageOutputStreamWriter writer)
			throws ProtocolException {
		super.serializeW(writer);
		try {
			writer.writeUint64(statement_id);
			writer.writeUint64(Statement_flagsEnum.toEnumMapValue(flags));
			writer.writeUint32(paramsCnt);
			if(params!=null)
				for (int i = 0; i < paramsCnt; i++)
					writer.writeVaruint(params[i]);
		} catch (Exception e) {
			throw new ProtocolException(e);
		}
	}

	@Override
	public boolean equals(Object obj) {
		if (!super.equals(obj))
			return false;
		Q_c_executePackage p = (Q_c_executePackage) obj;
		if (!(((statement_id == p.getStatement_id()) || ((statement_id != null)
				&& (p.getStatement_id() != null) && ((statement_id).equals(p
				.getStatement_id()))))))
			return false;
		if (!(((flags == p.getFlags()) || ((flags != null)
				&& (p.getFlags() != null) && ((flags).equals(p.getFlags()))))))
			return false;
		if (!((paramsCnt == p.getParamsCnt())))
			return false;
		
		if (params == p.getParams())
			return true;
		if (params==null || p.getParams()==null)
			return false;
		if (params.length!=p.getParams().length)
			return false;
		for (int i=0; i<paramsCnt; i++)
		{
			if ((p.getParams()[i]==params[i]))
				continue;
			if ((p.getParams()[i]==null)||(params[i]==null))
				return false;
			if (!(p.getParams()[i].equals(params[i])))
				return false;
		}
	
		return true;
	}

	// ================= GETTERS AND SETTERS ===================

	public java.math.BigInteger getStatement_id() {
		return statement_id;
	};

	public void setStatement_id(java.math.BigInteger a_statement_id) {
		a_statement_id = statement_id;
	};

	public EnumSet<Statement_flagsEnum> getFlags() {
		return flags;
	};

	public void setFlags(EnumSet<Statement_flagsEnum> a_flags) {
		a_flags = flags;
	};

	public long getParamsCnt() {
		return paramsCnt;
	};

	public void setParamsCnt(long a_paramsCnt) {
		a_paramsCnt = paramsCnt;
	};

	public Long[] getParams() {
		return params;
	};

	public void setParams(Long[] a_params) {
		a_params = params;
	};

}
